src/emu/layout/etibuffer.lay

```diff
@@ -0,0 +1,276 @@
+		if (not(pc >= 0x100 and pc <= 0x106) and -- avoid updating when pc is in these ranges due to EXX / EXX
+			not(pc >= 0x11a and pc <= 0x11c) and
+			not(pc >= 0x134 and pc <= 0x137) and
+			not(pc >= 0x159 and pc <= 0x15b)) then
+			bufferhead = file.device:subdevice("maincpu").state["BC"].value
+			buffertail = file.device:subdevice("maincpu").state["DE"].value
+			buffersize = file.device:subdevice("maincpu").state["HL"].value
+			local minsize = 512
+			if bufferhead >= buffertail then
+				bufferhead = max(buffertail + minsize, bufferhead)   -- minimum size or you can't see it
+				bufferfull = nil
+			else
+				bufferfull = iif (buffertail - bufferhead < minsize,1,nil)
+				buffertail = max(buffertail, bufferhead + minsize)  -- minimum size
+			end
+		end
+
+		bufheadx = scalevalue(bufferhead,0x4000,0xffff,boundx0,boundx1)
+		buftailx = scalevalue(buffertail,0x4000,0xffff,boundx0,boundx1)
+		if bufheadx >= buftailx then
+				middleboundx0 = buftailx
+				middleboundx1 = bufheadx
+				-- choose bar colors: middle bar white, full bar black
+				middlecolor, fullcolor = 0xffffff, 0x000000
+		else
+				middleboundx0 = bufheadx
+				middleboundx1 = buftailx
+				-- choose bar colors: full bar white   (reddish white when buffer full),
+				--                    middle bar black (red when buffer full)
+				middlecolor,fullcolor = iif(bufferfull, 0xff0000, 0x000000), iif(bufferfull, 0xffaaaa, 0xffffff)
+		end
+	end -- updatebufferdetails
+
+	-- led digits
+
+	for digit=0,3 do
+		file.views["Control Panel"].items["digit"..string.format("%d",digit)]:set_element_state_callback(
+			function ()
+				local pc = file.device:subdevice("maincpu").state["PC"].value
+				if (not(pc >= 0x100 and pc <= 0x106) and
+					not(pc >= 0x11a and pc <= 0x11c) and
+					not(pc >= 0x134 and pc <= 0x137) and
+					not(pc >= 0x159 and pc <= 0x15b))
+				then buffersize = file.device:subdevice("maincpu").state["HL"].value end
+				return ({ 0x3f, 0x06, 0x5b, 0x4f, 0x66, 0x6d, 0x7d, 0x07, 0x7f, 0x6f, 0x77, 0x7c, 0x39, 0x5e, 0x79, 0x71 })
+						[ (buffersize >> (digit * 4) & 0x0f) + 1 ]
+			end)
+	end
+
+	-- buffer head amd tail bars
+
+	file.views["Control Panel"].items["bufheadtaildisplay0"]:set_color_callback(
+		function ()
+			updatebufferdetails()
+			return rgbtorender_color(fullcolor)
+		end)
+
+	file.views["Control Panel"].items["bufheadtaildisplay0"]:set_bounds_callback(
+		function ()
+			updatebufferdetails()
+			local b = emu.render_bounds()
+			b:set_xy(boundx0,boundy0,boundx1,boundy1)
+			return b
+		end)
+
+	file.views["Control Panel"].items["bufheadtaildisplay1"]:set_color_callback(
+		function ()
+			updatebufferdetails()
+			return rgbtorender_color(middlecolor)
+		end)
+
+	file.views["Control Panel"].items["bufheadtaildisplay1"]:set_bounds_callback(
+		function ()
+			updatebufferdetails()
+			local b = emu.render_bounds()
+			b:set_xy(middleboundx0,boundy0,middleboundx1,boundy1)
+			return b
+		end)
+ end)
+
+]]></script>
```

Comment:
It still feels like you’re going out of your way to do things inefficiently in this layout:
* At least `bufheadtaildisplay0` is only animated in one dimension – you only need to provide the element state, and you can use the piecewise size/colour interpolation provided by the framework.
* You don’t need to update the buffer status every time you get a parameter callback, you only need to do it when preparing to draw a frame, and there’s a callback for that.
* You’re need to recalculate view coordinates on receiving the relevant callback.

Reply:
ok, I put bufferupdatedetails inside a set_prepare_items_callback:

	file.views["Control Panel"]:set_prepare_items_callback( -- called every frame
		function ()
			updatebufferdetails()
		end)

I'm not sure that the framework interpolation would work because there's two elements of state, the buffer head and the buffer tail and I thought the interpolation will only apply to a single element of state per layout object.  

(Oh how I wish I could just draw directly into the layout 8-)

So I have two bars, one is the full bar and there's a middle bar.

The full bar's bounds basically don't change, but the middle bar will extend from the buffer tail to the buffer head and if it wraps around, it exchanges the full bar's colors with the middle bar's colors so it appears to wrap around.

With regards to recalculating the view coordinates, I thought that was automatically done for you since the bounds coordinates are relative to the enclosing view?